File: baseble/src/main/java/com/fcbox/ble/baseble/callback/scan/SingleFilterScanCallback.java
package com.fcbox.ble.baseble.callback.scan;


import com.fcbox.ble.baseble.FcBoxBle;
import com.fcbox.ble.baseble.model.BluetoothLeDevice;

import java.util.concurrent.atomic.AtomicBoolean;


/**
 * 设置扫描指定的单个设备，一般是设备名称和 Mac 地址
 */
public class SingleFilterScanCallback extends ScanCallback {
    private AtomicBoolean hasFound = new AtomicBoolean(false);
    private String deviceName;//指定设备名称
    private String deviceMac;//指定设备Mac地址

    public SingleFilterScanCallback(IScanCallback scanCallback) {
        super(scanCallback);
    }

    public ScanCallback setDeviceName(String deviceName) {
        this.deviceName = deviceName;
        return this;
    }

    public ScanCallback setDeviceMac(String deviceMac) {
        this.deviceMac = deviceMac;
        return this;
    }

    @Override
    public BluetoothLeDevice onFilter(BluetoothLeDevice bluetoothLeDevice) {
        BluetoothLeDevice tempDevice = null;
        if (!hasFound.get()) {
            if (bluetoothLeDevice != null && bluetoothLeDevice.getAddress() != null && deviceMac != null
                    && deviceMac.equalsIgnoreCase(bluetoothLeDevice.getAddress().trim())) {
                hasFound.set(true);
                isScanning = false;
                removeHandlerMsg();
                FcBoxBle.getInstance().stopScan(SingleFilterScanCallback.this);
                tempDevice = bluetoothLeDevice;
                bluetoothLeDeviceStore.addDevice(bluetoothLeDevice);
                scanCallback.onScanFinish(bluetoothLeDeviceStore);
            } else if (bluetoothLeDevice != null && bluetoothLeDevice.getName() != null && deviceName != null
                    && deviceName.equalsIgnoreCase(bluetoothLeDevice.getName().trim())) {
                hasFound.set(true);
                isScanning = false;
                removeHandlerMsg();
                FcBoxBle.getInstance().stopScan(SingleFilterScanCallback.this);
                tempDevice = bluetoothLeDevice;
                bluetoothLeDeviceStore.addDevice(bluetoothLeDevice);
                scanCallback.onScanFinish(bluetoothLeDeviceStore);
            }
        }
        return tempDevice;
    }
}

File: baseble/src/main/java/com/xzy/ble/baseble/callback/scan/SingleFilterScanCallback.kt
package com.xzy.ble.baseble.callback.scan
import com.xzy.ble.baseble.BaseBle
import com.xzy.ble.baseble.model.BluetoothLeDevice

import java.util.concurrent.atomic.AtomicBoolean


/**
 * 设置扫描指定的单个设备，一般是设备名称和 Mac 地址
 */
@Suppress("unused")
class SingleFilterScanCallback(scanCallback: IScanCallback) : ScanCallback(scanCallback) {
    private val hasFound = AtomicBoolean(false)
    private var deviceName: String? = null//指定设备名称
    private var deviceMac: String? = null//指定设备Mac地址

    fun setDeviceName(deviceName: String): ScanCallback {
        this.deviceName = deviceName
        return this
    }

    fun setDeviceMac(deviceMac: String): ScanCallback {
        this.deviceMac = deviceMac
        return this
    }

    override fun onFilter(bluetoothLeDevice: BluetoothLeDevice): BluetoothLeDevice? {
        var tempDevice: BluetoothLeDevice? = null
        if (!hasFound.get()) {
            if (deviceMac != null && deviceMac!!.equals(bluetoothLeDevice.address.trim { it <= ' ' }, ignoreCase = true)) {
                hasFound.set(true)
                isScanning = false
                removeHandlerMsg()
                BaseBle.getInstance()?.stopScan(this@SingleFilterScanCallback)
                tempDevice = bluetoothLeDevice
                bluetoothLeDeviceStore.addDevice(bluetoothLeDevice)
                scanCallback!!.onScanFinish(bluetoothLeDeviceStore)
            } else if (deviceName != null && deviceName.equals(bluetoothLeDevice.name.trim { it <= ' ' }, ignoreCase = true)) {
                hasFound.set(true)
                isScanning = false
                removeHandlerMsg()
                BaseBle.getInstance()?.stopScan(this@SingleFilterScanCallback)
                tempDevice = bluetoothLeDevice
                bluetoothLeDeviceStore.addDevice(bluetoothLeDevice)
                scanCallback!!.onScanFinish(bluetoothLeDeviceStore)
            }
        }
        return tempDevice
    }
}
